File: validate_braces.py
from pathlib import Path
b = Path('shaders/fragment/boundaryShader.frag').read_text().splitlines()
c = Path('shaders/common.glsl').read_text().splitlines()
out=[]
for line in b:
    if line.strip() == '#include "common.glsl"':
        out.extend(c)
    else:
        out.append(line)
stack=[]
for i,line in enumerate(out, start=1):
    for ch in line:
        if ch == '{':
            stack.append((i,line))
        elif ch == '}':
            if stack:
                stack.pop()
            else:
                print('unmatched close', i, line)
print('unmatched opens', len(stack))
for lineno,line in stack:
    print('open at', lineno, line)

File: validate_braces.js
const fs = require('fs');
const b = fs.readFileSync('shaders/fragment/boundaryShader.frag','utf8').split(/\r?\n/);
const c = fs.readFileSync('shaders/common.glsl','utf8').split(/\r?\n/);
const out=[];
for(const line of b){ if(line.trim() === '#include "common.glsl"'){ out.push(...c); } else { out.push(line); } }
const stack=[];
for(let i=0;i<out.length;i++){
  const line=out[i];
  for(const ch of line){
    if(ch==='{' ) stack.push({line:i+1,text:line});
    else if(ch==='}'){
      if(stack.length) stack.pop();
      else console.log('unmatched close', i+1, line);
    }
  }
}
console.log('unmatched opens', stack.length);
for(const item of stack) console.log('open at', item.line, item.text);
